### PR TITLE
Set back to account confirmation on email change

### LINKED ISSUE #197: Updating email address in PATCH account should trigger a new confirmation process.
Otherwise, someone can enter in a valid email, and then edit their account so that they have an invalid email.

### PR DESCRIPTION
**Pull request by MrHarshPatel**
# Description

Now, updateAccount middleware changes whether the account is confirmed by finding an account and checking its email with the request.

updateAccount middleware then sets req.body.account to hold the changed account for sendConfirmAccountEmail.

The controller then deletes req.body.account and gives back req.body which holds only the values to be changed on a success (the ones provided).

Notes:
- /hacker/self now gives an Account 403 error on unconfirmed emails for hackers.
- If a hacker is unconfirmed they are not allowed to change their hacker info through /hacker/:id route. Similarly, they will not get emails (day-of, week-of etc.), their status and check in status cannot be changed even by admins until they reconfirm their email.

Fixes #197  (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Manually tested as well as running npm test (same test fail on develop will look into).
Also added tests.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: middlewares/account.middleware.js [hunk 1]
```diff
@@ -159,8 +159,16 @@ async function addAccount(req, res, next) {
  * @param {*} next 
  */
 async function updateAccount(req, res, next) {
-    const account = await Services.Account.updateOne(req.params.id, req.body);
-    if (account) {
+    var account = await Services.Account.findById(req.params.id)
+
+    // If we are changing the email, and there is a difference between the two, set back to unconfirmed status.
+    if (req.body.email && account.email != req.body.email) {
```
Comment (loreina) [inline middlewares/account.middleware.js]: why are we checking for both? is there a way we can not call this if `req.body.email` is unchanged?

File: routes/api/account.js [hunk 2]
```diff
@@ -253,6 +253,7 @@ module.exports = {
             Middleware.Account.parsePatch,
 
             Middleware.Account.updateAccount,
+            Middleware.Auth.sendConfirmAccountEmailMiddleware,
```
Comment (loreina) [inline routes/api/account.js]: can we change this so the word middleware isn't there twice? like to `Middleware.Auth.sendConfirmAccountEmail`? or maybe even shorter bc this is really long

File: middlewares/account.middleware.js (hunk 1 above)
Comment (MrHarshPatel) [reply middlewares/account.middleware.js]: Currently, if a field isn't given to the request then it doesn't change it. So if we don't check if req.body.email is undefined then if somebody has an email and they change their name, it'll make them unconfirmed.

Not sure of a better way

Comment (pierreTklein) [reply middlewares/account.middleware.js]: Would the PR for parsePatch break this check?

**Comment by pierreTklein** [comment]
Make sure you write some tests for the expected behavior caused by this change.

File: tests/hacker.test.js [hunk 3]
```diff
@@ -603,6 +651,30 @@ describe("PATCH update one hacker", function () {
         });
     });
 
+    // hacker should fail to checkin hacker due to unconfirmed email
+    it("should FAIL to check in hacker as a volunteer due to unconfirmed email", function (done) {
```
Comment (Ty-Won) [inline tests/hacker.test.js]: Hackers can be volunteers?

File: tests/hacker.test.js [hunk 4]
```diff
@@ -630,6 +702,30 @@ describe("PATCH update one hacker", function () {
         });
     });
 
+    // should fail to change hacker data with an unconfirmed email
+    it("should FAIL and FAIL to update the user's hacker info due to unconfirmed email", function (done) {
```
Comment (Ty-Won) [inline tests/hacker.test.js]: What does "Fail and Fail mean"?

File: tests/hacker.test.js (hunk 3 above)
Comment (loreina) [reply tests/hacker.test.js]: no it means a volunteer tried to check in a hacker

File: middlewares/auth.middleware.js [hunk 5]
```diff
@@ -455,7 +455,7 @@ module.exports = {
     parseResetToken: parseResetToken,
     validateResetToken: Middleware.Util.asyncMiddleware(validateResetToken),
     deleteResetToken: deleteResetToken,
-    sendConfirmAccountEmailMiddleware: Middleware.Util.asyncMiddleware(sendConfirmAccountEmailMiddleware),
+    sendConfirmAccountEmail: Middleware.Util.asyncMiddleware(sendConfirmAccountEmail),
```
[the remark this answers is not in the document]
Comment (MrHarshPatel) [reply middlewares/auth.middleware.js]: Hmm, seems to wrap the async function in another middleware and try to resolve the promise. If not resolved then returns error. See updateAccount in account.middleware.js for example.

File: tests/hacker.test.js (hunk 4 above)
Comment (MrHarshPatel) [reply tests/hacker.test.js]: changed to should FAIL and not update the user's hacker info...

File: middlewares/account.middleware.js (hunk 1 above)
Comment (MrHarshPatel) [reply middlewares/account.middleware.js]: Yes! Talked to @Ty-Won and he said to add TODO as his patch is currently breaking, for when this hits.

File: middlewares/hacker.middleware.js [hunk 6]
```diff
@@ -129,13 +129,11 @@ function addDefaultStatus(req, res, next) {
...
- * Verifies that account is confirmed and of proper type from the account ID passed in req.body.accountId
- * @param {{body: {accountId: ObjectId}}} req 
- * @param {*} res 
+ * Helper function that validates if account is confirmed and is of proper type
+ * @param account account object containing the information for an account 
  * @param {(err?) => void} next 
  */
-async function validateConfirmedStatus(req, res, next) {
-    const account = await Services.Account.findById(req.body.accountId);
+async function validateConfirmedStatus(account, next) {
```
Comment (Ty-Won) [inline middlewares/hacker.middleware.js]: does change in req, res, next structure affect how we chain the middleware?

Comment (Ty-Won) [reply middlewares/hacker.middleware.js]: also is this needed? if we put the if statement in validateConfirmedStatusFromAccountId does it still work?

Comment (MrHarshPatel) [reply middlewares/hacker.middleware.js]: Made them separated as we have 3 different validateConfirmed statuses: fromAccountID, fromHackerID and fromObject

**Pull request merged**